Add support for annotating queries generated by ActiveRecord::Relation with SQL comments

mattyoho opened the pull request:
#### Summary

This pull request implements a small new feature that allows users to annotate generated SQL queries.

A new method, `ActiveRecord::Relation#annotate`, has been added that allows annotating queries generated from the relation instance with block-style SQL comments.

For example:

```ruby
Post.annotate("this is a comment").where(id: 123).to_sql
# SELECT "posts".* FROM "posts" WHERE "posts"."id" = 123 /* this is a comment */
```

The hope is to land this feature in time for Rails 6.0. 🙂🤞

#### Discussion

This kind of annotation can be extremely useful when combined with instrumentation or other methods of tracking and analyzing queries generated at runtime, particularly in a large, mature application.

The [`marginalia` gem](https://github.com/basecamp/marginalia) provides similarly useful functionality in some contexts, but currently operates on all queries within a request or job and can't be used to target specific queries.

By hooking into `ActiveRecord::Relation` it's possible to annotate arbitrary scopes and model associations.

```ruby
class Post < ActiveRecord::Base
  has_many :tags, -> { annotate("association comment") }

  scope :published, -> { where.not(published_at: nil).annotate("scope comment") }
end
```

This PR also updates `ActiveRecord::Relation#update_all` and `ActiveRecord::Relation#delete_all` to respect any annotations on the receiving relation object.

Multiple annotations may be added:

```ruby
Post.where(id: 123).with_annotation("foo").annotate("bar")
# SELECT "posts".* FROM "posts" WHERE "posts"."id" = 123 /* foo */ /* bar */
```

Not all queries issued from `ActiveRecord` can be annotated using this approach,  but many can.

#### Implementation

The implementation extends `Arel` under the hood by adding a new `Arel::Nodes::Comment` node and associated plumbing to work with them.

~Of note, `Arel::Nodes::Comment` is responsible for preventing SQL injection via comment escaping.~

~Comment content is scrubbed of SQL block comment delimiters at creation time [here](https://github.com/rails/rails/blob/6d85bc596389ba0118fb90e843d1abc2a68fea6e/activerecord/lib/arel/nodes/annotation.rb#L39). This should be reviewed for correctness. :eyes:~

#### Example use with instrumentation

Here's a toy example of how this could be used with `ActiveSupport::Notifications` instrumentation to perform basic analysis.

```ruby
query_checker = Proc.new do |_, _, _, _, payload|
  query = payload[:sql]
  if query =~ /JOIN "forbidden_table"/ && query !~ %r{/\* whitelisted \*/}
    Rails.logger.warn("this is bad!")
  end
end
ActiveSupport::Notifications.subscribe "sql.active_record", query_logger

# This would be all right
Post.annotate("whitelisted").joins(:forbidden_table).first
# But this would raise eyebrows
Post.joins(:forbidden_table).first
```

### Rails Guides

It might be useful to add documentation for this new method to the [Active Record Query Interface](https://guides.rubyonrails.org/active_record_querying.html) page.

I can provide a separate PR for that or add it here if so.

mattyoho committed:
Add Relation#annotate for SQL commenting

This patch has two main portions:

1. Add SQL comment support to Arel via Arel::Nodes::Comment.
2. Implement a Relation#annotate method on top of that.

== Adding SQL comment support

Adds a new Arel::Nodes::Comment node that represents an optional SQL
comment and teachers the relevant visitors how to handle it.

Comment nodes may be added to the basic CRUD statement nodes and set
through any of the four (Select|Insert|Update|Delete)Manager objects.

For example:

    manager = Arel::UpdateManager.new
    manager.table table
    manager.comment("annotation")
    manager.to_sql # UPDATE "users" /* annotation */

This new node type will be used by ActiveRecord::Relation to enable
query annotation via SQL comments.

== Implementing the Relation#annotate method

Implements `ActiveRecord::Relation#annotate`, which accepts a comment
string that will be appeneded to any queries generated by the relation.

Some examples:

    relation = Post.where(id: 123).annotate("metadata string")
    relation.first
    # SELECT "posts".* FROM "posts" WHERE "posts"."id" = 123
    # LIMIT 1 /* metadata string */

    class Tag < ActiveRecord::Base
      scope :foo_annotated, -> { annotate("foo") }
    end
    Tag.foo_annotated.annotate("bar").first
    # SELECT "tags".* FROM "tags" LIMIT 1 /* foo */ /* bar */

Also wires up the plumbing so this works with `#update_all` and
`#delete_all` as well.

This feature is useful for instrumentation and general analysis of
queries generated at runtime.

rafaelfranca commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -981,6 +995,12 @@ def build_arel(aliases)
         arel.from(build_from) unless from_clause.empty?
         arel.lock(lock_value) if lock_value
 
+        with_annotation_values.reject(&:nil?).tap do |annotations|
```
Why reject `nil?`? Why would someone pass `nil` for that method?

mattyoho replied:
This was an artifact of iterating on the implementation, I think. Good points, thank you. I've simplified things in the latest version.

rafaelfranca commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -981,6 +995,12 @@ def build_arel(aliases)
         arel.from(build_from) unless from_clause.empty?
         arel.lock(lock_value) if lock_value
 
+        with_annotation_values.reject(&:nil?).tap do |annotations|
```
`tap` here is confusing an it would be better to just assign to a variable since the value itself is not returned.

jeremy commented on activerecord/lib/arel/nodes/select_core.rb:
```diff
@@ -55,7 +57,8 @@ def eql?(other)
+          self.comment == other.comment
       end
       alias :== :eql?
     end
```
Suggests there's a single comment per statement, when there's a single annotation that is a comment node.

What would we call this if we had other comments in the statement? Annotation as well?

jeremy commented on activerecord/lib/arel/nodes/comment.rb:
```diff
@@ -0,0 +1,41 @@
+            str = str.gsub(SQL_COMMENT_DELIMITER_PATTERN, "")
+          end
+          str
+        end
```
Should we catch magic comment injection here as well, e.g. `/*! … ` extensions in MySQL or `/*+` optimizer hints?

We do prevent these cases due to the SQL visitor emitting spaces around the value. Would we notice the injection if it stopped emitting spaces? (And whose responsibility is it to recognize that: the SQL visitor or the node?)

jeremy commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,16 @@
+    This can be useful in instrumentation or other analysis of issued queries.
+
+    *Matt Yoho*
+
```
The `with_` feels mismatched with other relation builder methods, suggesting e.g. an annotation that applies within the scope of a block. Analog would be `with_conditions` vs `where` or `with_limit` vs `limit`. Would compare with `Post.where(id: 123).annotate("this is a comment")` or similar.

jeremy commented on activerecord/test/cases/arel/nodes/comment_test.rb:
```diff
@@ -0,0 +1,40 @@
+          compile(node).must_be_like %{ /* foo */ }
+
+          node = Comment.new("**//foo//**")
+          compile(node).must_be_like %{ /* foo */ }
```
Nice test coverage double delimiters. Do `#`/`--` comments interfere with inline comments at all?

jeremy reviewed (commented):
Welcome addition and great, thorough PR. Thank you!

mattyoho reviewed (commented):
Thanks for the code review, @jeremy!

I've switched to adding distinct commits while I'm iterating based on review feedback. I can squash back down toward the end.

The **Support Optimizer Hints** PR in https://github.com/rails/rails/pull/35615 is really interesting in relation to this one. **Edit**: Or vice-versa. 😄

From https://github.com/rails/rails/pull/35615#discussion_r265821777:
> Should we ensure the hint is safe for a SQL comment, preventing query injection?
>
> Could reuse the Arel Comment node #35617

I do think there'd be value in some slight generalization here to allow https://github.com/rails/rails/pull/35615 to reuse some of this. I've left some comments inline discussing it more.

In particular in this PR since the last review:

- https://github.com/rails/rails/pull/35617/commits/4a4003aad2bccc8b2be664ee1e90459edb83a651 renames `Arel::Nodes::Comment` to `Arel::Nodes::Annotation`.
- https://github.com/rails/rails/pull/35617/commits/acdd31616ce026241f4b8db44f50d5000627f4ab demonstrates `Arel::Nodes::Annotation` representing an optimization hint within the visitor.

mattyoho commented:
Howdy. :wave:

I've iterated based on some review feedback and self-review, added some additional tests, and incorporated the latest changes from `master`, including https://github.com/rails/rails/pull/35615.

I think this patch is now in a state where it could be merged and so is ready for review again. 🙂🤞

I think it's worth me pointing out that I've refactored the additions in https://github.com/rails/rails/pull/35615 a bit after incorporating them into this branch.

In particular, I've changed `Arel::Nodes::OptimizerHints` to inherit from `Arel::Nodes::Annotation` rather than `Arel::Nodes::Unary`.

I think this makes conceptual sense, but the biggest practical advantage comes from reusing the comment sanitization routine rather than having two adjacent implementations.

You can see the changes specific to this refactor isolated in [this Gist](https://gist.github.com/mattyoho/456cdd5668c30020933b1a55bc991443).

Thank you!

jeremy commented on activerecord/test/cases/relation_test.rb:
```diff
@@ -307,6 +307,63 @@ def test_relation_merging_keeps_joining_order
+      assert_sql(%r{/\* foo bar \*/}) do
+        assert post_with_annotation.first, "record should be found"
+      end
+    end
```
Wrapping my head around annotate with multiple arguments. Joining with a space in the SQL statement feels surprising, and I'm not sure what's meant by more than one annotation. For multivalue, I'd imagine we'd want a standard format and key/value pairs rather than arbitrary strings.

mattyoho replied:
> Joining with a space in the SQL statement feels surprising, and I'm not sure what's meant by more than one annotation.

Yeah, that's a great point. I suppose I've done this because I could, not because I should. I'll walk it back.

> For multivalue, I'd imagine we'd want a standard format and key/value pairs rather than arbitrary strings.

Yeah, I'd considered this, i.e., accepting a hash and generating `"key1:val1,key2:val2"` strings, but that also seems like something that can be handled as per developer preferences at the application level.

jeremy commented on activerecord/lib/arel/insert_manager.rb:
```diff
@@ -38,6 +38,11 @@ def insert(fields)
+      @ast.comment = Nodes::Annotation.new(*values)
+      self
+    end
+
```
Seems the AST node is an `annotation`, since it's a specific usage of a comment that always appears at the end of the query.

mattyoho replied:
I see what you're saying. Perhaps if `Annotation` is the mixin and we're instantiating a `Comment` here then this property name `comment` makes sense.

mattyoho replied:
(At least, I _think_ I see what you're saying. 🙂 If it seems like I've misinterpreted you please let me know.)

Something that may be clear by inference but I'll make explicit, the naming at the `AR::Relation` level around "annotate" / "annotation" was chosen primarily because of the likely number of `:comment` / `:comments` associations and scopes in the wild, and wanting to avoid ambiguity or collision there.

At the Arel AST level, in terms of the SQL domain, the term "comment" seems the most precise. Hence the intent to switch back to a `Comment` concrete node class and name the more generic mixin `Annotation`.

mattyoho replied:
As of the latest code, there are now `Arel::Nodes::Comment` and `Arel::Nodes::OptimizerHints` classes that each mix in `Arel::Nodes::Annotation`. I think that naming makes the most sense in the SQL domain, but if you disagree I can change it around.

jeremy commented on activerecord/lib/arel/nodes/annotation.rb:
```diff
@@ -0,0 +1,35 @@
+
+module Arel # :nodoc: all
+  module Nodes
+    class Annotation < Arel::Nodes::Node
```
Optimizer hints and annotation nodes are both serialized as comments in SQL, thanks to the SQL visitor, but as Arel nodes, does OptimizerHints have an _is-a_ relationship with Annotation? Consider a `Comment` mixin (or superclass?) that clarifies that their relationship is implementation only.

Furthermore, should comment sanitization be the node's responsibility, or the SQL visitor's?

/cc @kamipo

mattyoho replied:
I think in my mental model, `Annotation` was the superclass, representing a non-executable (hand-wavey term) piece of metadata added to the query. Maybe `Annotation`, which handles comment sanitization, is the mixin and `OptimizerHints` and `Comment` are the concrete classes.

> Furthermore, should comment sanitization be the node's responsibility, or the SQL visitor's?

I do think SQL comment sanitization should be done at the node level.

At the node level it happens once, up front, at node creation time, rather than having to be repeated through multiple visitor methods. If a relation is used to generate multiple queries, the sanitization cost is paid once also.

All types of annotations have to care about SQL comment injection, so I think it makes sense to centralize it.


From https://github.com/rails/rails/pull/35615#pullrequestreview-215377223:
> Does this need different sanitization than `/* */` removal? Consider the hint `) …other SQL clauses… OPTION(`.

That would result in a query error, as multiple `OPTION` clauses in a statement is invalid. (Doc [here](https://docs.microsoft.com/en-us/sql/t-sql/queries/option-clause-transact-sql?view=sql-server-2017).)

I think the risk profile there is much lower and probably there's no need to sanitize it. (Sanitization would probably just result in a differently invalid query.) But if it did need to be escaped, that seems like something to handle in the database-specific visitor.

mattyoho reviewed (commented):
I've rebased off latest `master` and attempted to respond to the most recent feedback. 

The `#annotate` query method now expects a `String` or `Arel::Nodes::SqlLiteral` argument.

Sanitization of SQL block comment delimiters is still happening at the node level, within the `Arel::Nodes::Annotation` mixin, for the reasons I proposed in https://github.com/rails/rails/pull/35617#discussion_r266265556, but I'm of course willing to incorporate further feedback on that based on the discussion. 🙂

mattyoho commented:
> Sanitization of SQL block comment delimiters is still happening at the node level, within the Arel::Nodes::Annotation mixin, for the reasons I proposed in #35617 (comment), but I'm of course willing to incorporate further feedback on that based on the discussion.

When https://github.com/rails/rails/pull/35660 is merged, I'll update this diff to reuse that approach. :+1:

mattyoho commented:
The latest base branch changes have been incorporated and sanitization in the visitors has been embraced. I think this is ready for review again. /cc @jeremy

eileencodes reviewed (approved):
Looks good to me. @jeremy anymore concerns or do you think this is good to merge?

jeremy commented:
lgtm!
On Thu, Mar 21, 2019 at 00:59 Eileen M. Uchitelle <notifications@github.com>
wrote:

> *@eileencodes* approved this pull request.
>
> Looks good to me. @jeremy <https://github.com/jeremy> anymore concerns or
> do you think this is good to merge?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/35617#pullrequestreview-216818879>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAAxwhyOlNQTCbBwPaN_F5dkJxVt51bks5vYlrGgaJpZM4b1fam>
> .
>

kamipo commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -5,7 +5,7 @@ module ActiveRecord
     MULTI_VALUE_METHODS  = [:includes, :eager_load, :preload, :select, :group,
                             :order, :joins, :left_outer_joins, :references,
-                            :extending, :unscope, :optimizer_hints]
+                            :extending, :unscope, :optimizer_hints, :annotate]
```
`:annotate` is no longer multi value.

kamipo commented on activerecord/lib/active_record/relation/query_methods.rb:
```diff
@@ -948,6 +948,25 @@ def skip_preloading! # :nodoc:
+    #   # SELECT "users"."name" FROM "users" /* selecting user names */
+    #
+    # The SQL block comment delimiters, "/*" and "*/", will be added automatically.
+    def annotate(comment)
```
This looks odd to me. All multi value query methods defined by `def multi_value_method(*args)` except the `annotate`.

mattyoho replied:
This is a little outside the norm for multi-value methods, but I believe this is a necessary compromise between the [desired](https://github.com/rails/rails/pull/35617#discussion_r266260622) API for `annotate` and the requirements of working with [`Relation::Merger`](https://github.com/rails/rails/blob/9c1adcaaf7ec27b5f994fe119c8a56516a514bea/activerecord/lib/active_record/relation/merger.rb#L76), e.g., to pass [this test](https://github.com/rails/rails/blob/9c1adcaaf7ec27b5f994fe119c8a56516a514bea/activerecord/test/cases/relation/merging_test.rb#L139).

kamipo replied:
What is the desired behavior for `User.annotate("foo").annotate("bar")`?

mattyoho replied:
The desired result would be a single SQL comment, `/* foo bar */`.

This is inline with how `#optimizer_hints` includes each hint in a single block, though of course that's required by the way hints are implemented across databases.

The alternative would be `/* foo */ /* bar */`. Do you think that is preferable?

kamipo replied:
Would be a single SQL comment is fine for now (merging different relation comments `Post.annotate("post comment").joins(:tags).merge(Tag.annotate("tag comment")) #=> "/* post comment tag comment */"` looks odd to me though).

My point is that, if defined by `def annotate(comment)`, it reminds me of single value method (i.e. `User.annotate("foo").annotate("bar") # => /* bar */`, that is the reason for the my confusion https://github.com/rails/rails/pull/35617#discussion_r267547547).

Since the `def annotate(comment)` makes any other confusion, it would be better to allow multiple values as multiple values query method like other multiple values query methods (`select`, `group`, etc.) for consistency for now.

kamipo replied:
I'm fine with `User.annotate("foo").annotate("bar") # => /* foo bar */`.

But if it behaves so, the single value method like method definition makes me (and any other people) confused that the `annotate` is single value method (i.e. `User.annotate("foo").annotate("bar") # => /* bar */`, that is the reason for the my confusion https://github.com/rails/rails/pull/35617#discussion_r267547547).

As a multiple values method, the `select` behaves `User.select("foo").select("bar") # => SELECT "users"."foo", "users"."bar" ...`, so as the same multiple values method, it is worth to be defined the `annotate` in the same way.

Can you change it back to be defined in the multiple values method way?

cc @jeremy

kamipo commented on activerecord/lib/arel/nodes/optimizer_hints.rb:
```diff
@@ -0,0 +1,19 @@
+module Arel # :nodoc: all
+  module Nodes
+    class OptimizerHints < Arel::Nodes::Node
+      include Arel::Nodes::Annotation
```
There is no worth to share `Arel::Nodes::Annotation` anymore.

mattyoho reviewed (commented):
> There is no worth to share `Arel::Nodes::Annotation` anymore.

Removed with latest push.

kamipo commented on activerecord/lib/arel/nodes/optimizer_hints.rb:
```diff
@@ -0,0 +1,29 @@
+
+module Arel # :nodoc: all
+  module Nodes
+    class OptimizerHints < Arel::Nodes::Node
```
There is no reason to touch the `OptimizerHints` node in this PR anymore.

mattyoho replied:
`OptimizerHints` is no longer impacted by this PR. :+1:

kamipo commented on activerecord/lib/arel/nodes/comment.rb:
```diff
@@ -0,0 +1,29 @@
+    class Comment < Arel::Nodes::Node
+      attr_reader :values
+
+      def initialize(*values)
```
I'd not prefer the splat args for the internal node, this allows `Comment.new(single_value)` even the `values` expects multiple values, but that is not necessary flexibility for internal nodes.

mattyoho replied:
This has been removed. :+1:

kamipo commented on activerecord/test/cases/adapters/mysql2/annotate_test.rb:
```diff
@@ -0,0 +1,37 @@
+require "cases/helper"
+require "models/post"
+
+class Mysql2AnnotateTest < ActiveRecord::Mysql2TestCase
```
Can you consolidate the duplicated `test/cases/adapters/*/annotate_test.rb`?
Those are not adapter specific tests.

kamipo replied:
ah I see, that is for the quoting things for `assert_sql`, it is fine as is for now.

kamipo commented on activerecord/lib/arel/table.rb:
```diff
@@ -78,6 +78,10 @@ def having(expr)
       from.having expr
     end
 
+    def comment(*cmt)
```
There is no reason to add the unused internal API.

mattyoho replied:
This was added because some developers do use `Arel` directly in applications. It has been removed. :+1:

kamipo replied:
Still remained.

kamipo commented on activerecord/lib/arel/visitors/to_sql.rb:
```diff
@@ -162,14 +164,19 @@ def visit_Arel_Nodes_SelectCore(o, collector)
         end
 
+        def visit_Arel_Nodes_Comment(o, collector)
+          comments = o.values.map { |v| sanitize_as_sql_comment(v) }.join(" ")
```
How about joining multiple comments with `"; "`?

I think it is one solution to mitigate the issue `Post.annotate("post comment").joins(:tags).merge(Tag.annotate("tag comment")) #=> "/* post comment tag comment */"`.

mattyoho replied:
I don't think text content should be inserted into the user's annotation if it can be avoided. As a consumer of the API I wouldn't want that.

After more reflection I think the best approach by far is actually to wrap each annotation in its own delimiters, e.g.,

```
Post.annotate("post comment").joins(:tags).merge(Tag.annotate("tag comment"))
#=> "... /* post comment */ /* tag comment */"
``` 

This solves the concerns with running them together, it makes it much simpler to detect or parse out annotation in instrumentation, and it prevents different developers from stepping on one another's annotations.

I've been using a "backport" of this API in an experimental branch for a large refactor and after working with it a bit this is the behavior I'd want to have.

kamipo commented on activerecord/test/cases/relation_test.rb:
```diff
@@ -307,6 +307,58 @@ def test_relation_merging_keeps_joining_order
+    end
+
+    def test_relation_without_annotation_does_not_include_an_empty_comment
+      log = assert_sql do
```
```suggestion
      log = capture_sql do
```

kamipo commented on activerecord/test/cases/scoping/relation_scoping_test.rb:
```diff
@@ -130,6 +130,44 @@ def test_scoped_count
+  def test_find_with_annotation_unscoped
+    Developer.annotate("scoped").unscoped do
+      developer = nil
+      log = assert_sql do
```
```suggestion
      log = capture_sql do
```

kamipo commented on activerecord/test/cases/scoping/relation_scoping_test.rb:
```diff
@@ -130,6 +130,44 @@ def test_scoped_count
+
+  def test_find_with_annotation_unscope
+    developer = nil
+    log = assert_sql do
```
```suggestion
    log = capture_sql do
```

kamipo commented on activerecord/lib/arel/insert_manager.rb:
```diff
@@ -38,6 +38,11 @@ def insert(fields)
       self
     end
 
+    def comment(*values)
```
Is this internal API necessary (no way to use it by public API (Active Record))?

mattyoho commented:
@kamipi Thank you for the thorough code review. The patch is much better because of it.

I believe this PR is once again ready for review and hopefully merging.

arthurnn commented:
On marginalia we do some escaping on the comment, to make sure people will not annotate queries with user input and be open to SQL injections.
Is that necessary in here? (I scammed the code fast, and haven't saw any prevention like that)

mattyoho commented:
@arthurnn That's being handle here: https://github.com/rails/rails/blob/f41825809cf3d27f98467bd7bc005e5d6ed94828/activerecord/lib/arel/visitors/to_sql.rb#L176

(Or [here](https://github.com/rails/rails/pull/35617/files#diff-eaaebca1a5b748881c9e793f50b76fdfR176) in this diff.)

mattyoho commented:
(There are tests, e.g, [here](https://github.com/rails/rails/pull/35617/files#diff-fa3aafa2872c56f3a7faed1058c9f22fR16) and [here](https://github.com/rails/rails/pull/35617/files#diff-8ea9ba8004ec342431443f6cc4bae152R16), as well.)

abhaynikam commented on guides/source/active_record_querying.md:
```diff
@@ -59,6 +59,7 @@ To retrieve objects from the database, Active Record provides several finder met
 
 The methods are:
 
+* `annotate`
```
Is `annotate` a finder method? Just curious to understand this because the documentation on the above line says 

```
To retrieve objects from the database, Active Record provides several finder methods. Each finder method allows you to pass arguments into it to perform certain queries on your database without writing raw SQL.
``` 

As per the changelog, it says `Add 'ActiveRecord::Relation#annotate' for adding SQL comments to its queries.`

mattyoho replied:
Hi, @abhaynikam! It's possible the description should be generalized a bit. Several other relation query methods arguably don't fit that documention sentence literally, either — `preload`, `create_with`, and `includes` being examples.

Perhaps the intro sentence could be updated in another PR.

abhaynikam replied:
Thanks, @mattyoho .  That makes sense, I agree a few other methods do not fit the description. 

Great addition @mattyoho 🎉

abhaynikam replied:
About updating the description, I would say let remove methods which do not fit description sinces that particular documentation section is for understanding how to retrieve objects from database. 

I'll like to raise a PR if somebody from Rails team approve it is correct to update the doc 😊
